// libcardano/blockchain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockChainBase = void 0;
class BlockChainBase {
    constructor() {
        this.subscribers = {
            'extendHeader': [],
            'rollback': [],
            'extendBlock': []
        };
        this.pipelines = {
            'extendHeader': [],
            'rollback': [],
            'extendBlock': []
        };
    }
    unsubscribe(event, handler) {
        const subscriberList = this.subscribers[event];
        for (let i = 0; i < subscriberList.length; i++) {
            if (subscriberList[i] == handler) {
                subscriberList.splice(i, 1);
            }
        }
    }
    setBlockSource(source) {
        this.blockSource = source;
    }
    notifyEvent(eventType, cb, ...args) {
        let counter = this.pipelines[eventType].length;
        const callback = (err) => {
            if (err) {
                counter = -1;
                cb(err);
            }
            else {
                counter--;
                if (counter == 0) {
                    this.subscribers[eventType].forEach((sus) => {
                        sus(...args);
                    });
                    cb();
                }
            }
        };
        if (this.pipelines[eventType].length == 0) {
            setImmediate(cb);
        }
        else {
            this.pipelines[eventType].forEach((sus) => {
                sus(...args, callback);
            });
        }
    }
    on(event, handler) {
        this.subscribers[event].push(handler);
    }
    pipeline(event, handler) {
        this.pipelines[event].push(handler);
    }
}
exports.BlockChainBase = BlockChainBase;
